src/swbf_fnv_hashes.cpp

```diff
@@ -1,2834 +1,3935 @@
-   "WeaponShareAmmo"_fnvp,
-   "WeaponShareHeat1"_fnvp,
-   "WeaponShareHeat2"_fnvp,
-   "WeaponShareHeat3"_fnvp,
-   "WeaponShareHeat4"_fnvp,
-   "WeaponShareHeat"_fnvp,
-   "WeaponUnableSelect"_fnvp,
-   "WheelOmegaToU"_fnvp,
-   "WheelOmegaToV"_fnvp,
-   "WheelSection"_fnvp,
-   "WheelTexture"_fnvp,
-   "WheelVelocToU"_fnvp,
-   "WheelVelocToV"_fnvp,
-   "Widescreen"_fnvp,
-   "Width"_fnvp,
-   "WiggleAmount"_fnvp,
-   "WiggleSpeed"_fnvp,
-   "win"_fnvp,
-   "winddirection"_fnvp,
-   "WindDownSound"_fnvp,
-   "WindSound"_fnvp,
-   "windspeed"_fnvp,
-   "WindUpSound"_fnvp,
-   "WingModel"_fnvp,
-   "word"_fnvp,
-   "word_reformat"_fnvp,
-   "WorldAligned"_fnvp,
-   "WorldBrightness"_fnvp,
-   "WorldContrast"_fnvp,
-   "WorldFogRange"_fnvp,
-   "WorldHeight"_fnvp,
-   "WorldRange"_fnvp,
-   "WorldSaturation"_fnvp,
-   "WrapInput"_fnvp,
-   "YawLimits"_fnvp,
-   "YawSpread"_fnvp,
-   "YawTurnFactor"_fnvp,
-   "YOffset"_fnvp,
-   "ZeroGravity"_fnvp,
-   "ZoomFirstPerson"_fnvp,
-   "ZoomMax"_fnvp,
-   "ZoomMin"_fnvp,
-   "ZoomRate"_fnvp,
-   "ZoomTurnDivisorMax"_fnvp,
-   "ZoomTurnDivisorMin"_fnvp,
-   "ZOrder"_fnvp};
-}
+std::mutex dictionary_mutex;
 
 std::string lookup_fnv_hash(const std::uint32_t hash)
 {
+   std::string retVal;
+
+   dictionary_mutex.lock();
    const auto result = swbf_hashes.find(hash);
 
-   if (result != std::cend(swbf_hashes)) return std::string{result->second};
+   if (result != std::cend(swbf_hashes)) {
+      // this can be interesting when building a dict
+      //synced_cout::print("looked_up:", result->second, "\n");
+      retVal = std::string{result->second};
+   }
+   dictionary_mutex.unlock();
+
+   if (retVal.empty()) {
+      // here hex is nice for 2 reasons:
+      //   1. easier to see in hex editor
+      //   2. it matches the output of the ToolsFl\bin\hash.exe program
+      char hex_string[20];
+      sprintf(hex_string, "0x%x", hash);
+      synced_cout::print("Warning: Unknown hash looked up.\n"s, "   value: "s, hex_string,
+                         '\n');
+      retVal = std::string(hex_string);
+   }
+   return retVal;
+}
+
+// Read in a file of strings that can be resolved by the hash lookup
+void read_dictionary(const char* fileName)
+{
+   synced_cout::print("Reading dictionary: "s, fileName, '\n');
+   std::ifstream input(fileName);
+   for (std::string line; getline(input, line);) {
+      add_hash(line);
+   }
+   input.close();
+   initialize_internal_dictionary();
+}
 
-   synced_cout::print("Warning: Unknown hash looked up.\n"s, "   value: "s, hash, '\n');
+// adds an entry to the string hashes
+// returns true if it was added ()
+bool add_hash(std::string str)
+{
+   bool retVal = false;
+   if ("" != str) {
+      uint32_t hash = fnv_1a_hash(str);
+      dictionary_mutex.lock();
+      if (swbf_hashes.find(hash) == std::cend(swbf_hashes)) {
+         swbf_hashes[hash] = str;
+         retVal = true;
+      }
+      dictionary_mutex.unlock();
+   }
+   return retVal;
```

Comment:
Using [std::lock_guard](https://en.cppreference.com/w/cpp/thread/lock_guard) (which is like `std::shared_lock` except it blocks until it has sole ownership of the mutex, which we need in order to safely write to `swbf_hashes`)  and [structured bindings](https://en.cppreference.com/w/cpp/language/structured_binding) ([this](https://www.youtube.com/watch?v=aBZlbb9sE-g) is probably a better place to learn them if you're unfamiliar and curious) we can vastly simplify this function, making it both safer and easier to read.


```suggestion
   std::lock_guard lock{swbf_hashes_mutex};

   const auto [element_iterator, inserted] = swbf_hashes.emplace(fnv_1a_hash(str), str);

   return inserted;
```

Reply:
Please review mutex handling to ensure I understood your suggestion here.